_posts/pl/2022-02-10-porownanie-jezykow-cloud-functions.md

```diff
@@ -0,0 +1,147 @@
+
+Z tego powodu postanowiłem wykonać test ponownie, aby wykluczyć różnice w czasie (ponieważ o godzinie 20:00 czasu Polskiego, w Hong Kongu była godzina 03:00). Dzięki temu mogłem sprawdzić, czy wpływ na wyniki ma tutaj fakt, że w środku nocy obciążenie centrum danych może być mniejsze.
+
+Drugi test wykluczył jednak kwestie godziny w danym regionie, ponieważ i tym razem Azja wypadła najlepiej. Z powodu odległości można zaobserwować znacznie wyższe minimalne i maksymalne czasy odpowiedzi, jednak średnio były one i tak nieco niższe niż w przypadku Frankfurtu, dzięki czemu w ciągu 10 minut udało się wykonać sumarycznie więcej żądań.
+
+Najgorzej wypadł tutaj region w USA, gdzie pomimo niższych minimalnych czasów odpowiedzi, średnio były one znacznie wyższe (co idealnie pokazuje kolumna z 95 percentylem). W efekcie funkcje osadzone w USA obsłużyły zauważalnie mniejszą liczbę żądań.
+
+Testy starałem się wykonać w środku tygodnia, aby były jak najbardziej miarodajne. Pod uwagę należy wziąć jednak fakt, że funkcje były bardzo prymitywne - jedyne co robiły, to odpowiadały "Hello World". Całkiem możliwe, że w przypadku funkcji do których wysyłamy lub które zwracają jakieś dane wyniki byłyby zupełnie inne, jednak zależało mi na sprawdzeniu prostych funkcji, ponieważ w tym przypadku łatwo jest porównać środowiska uruchomieniowe (w przypadku cięższych implementacji duży wpływ na wydajność mogłyby mieć wykorzystane zewnętrzne zależności czy biblioteki).
+
+Podsumowując, gdybym chciał osadzić prostą funkcję i zależałoby mi, aby obsłużyła jak największy ruch, prawdopodobnie wybrałbym któryś z regionów w Azji.
+
+### Środowiska uruchomieniowe
+
+W przypadku środowisk uruchomieniowych spodziewałem się, że języki interpretowane dają lepsze wyniki niż języki kompilowane (czy uruchamiane w wirtualnej maszynie).
+
+Wyniki testu potwierdziły moje podejrzenia, ponieważ najszybciej odpowiadały funkcje napisane w Go, Ruby czy PHP.
+
+Kompletnie nie zdziwiły mnie wyniki funkcji napisanych w Javie czy .NET, jednak nie spisywałbym ich na straty. Środowiska uruchomieniowe wykorzystujące maszyny wirtualne (takie jak właśnie Java - JVM, czy .NET - CLR) potrafią optymalizować uruchomiony kod, jednak nie zrobią tego od razu, ponieważ potrzebują w tym celu zebrać odpowiednią ilość statystyk. Całkiem możliwe, że funkcje które obsługują bardzo dużo żądań w czasie ciągłym (czyli takim, dzięki któremu instancja funkcji będzie żyła bardzo długo) osiągnęłyby z czasem lepsze wyniki.
+
+Jakie z tego wnioski? Jeżeli piszemy prostą funkcję i nie zależy nam na wydajności (albo spodziewamy się małego ruchu), śmiało możemy napisać ją w języku, który znamy najlepiej. Jeżeli jednak zależy nam na obsłudze jak największej ilości żądań (i jednocześnie wiemy że instancja funkcji nie będzie długowieczna), najlepszym wyborem będą języki interpretowane/skryptowe.
```

Comment:
Zastanawiam się, czy nie warto skomentować dość przeciętnych wyników JS... Tymbardziej, że często pojawia się taki argument, ze node.js jest najlepszą platformą dla Functions

Reply:
https://medium.com/google-cloud/serverless-performance-comparison-does-the-language-matter-c72a7191c799
Tutaj porównanie języków w Cloud Functions i Cloud Run, cytując:

> NodeJS results vary from simple to double. 1 time fast and 2 times slow with Cloud Functions, and 2 times fast and 1 time slow with Cloud Run.
I ran the tests several times and I always got this result. I’m unable to explain it, especially because Nodejs is my weaker language among the 4. Please suggest!

Reply:
Dodam komentarz o zaskoczeniu wynikami Node.js, ale niestety nie mam na to żadnych wniosków i nie mam pojęcia czym to jest spowodowane. Node.js wypadł bliżej Javy niż języków interpretowanych jak Go czy PHP :/